---
title: "Codon Usage"
author: "Abhishek Jain"
date: "18 June 2019"
output: 
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
---

## Data exploration

```{r setup}

if(!require(tidyverse)){
    install.packages("tidyverse")
    library(tidyverse)
}
if(!require(glmnet)){
    install.packages("glmnet")
    library(glmnet)
}
if(!require(ggplot2)){
    install.packages("ggplot2")
    library(tidyverse)
}

```

```{R Load Data}

ref_raw <- read_rds("data/Sun_mutation_UTRs.rds")
  #Get sequences from ref_raw in a separate vector
  cds_seq <- ref_raw$CDS_seq
  
sc_raw <-read_tsv("data/YPD_scRNA.tab")

#Load Manually created motifs list into a vector
#motifs_raw <- scan("data/list_motifs.txt", character())
#list_genes <- scan("data/genes_terminators_list.txt", character())

#motifs_cheng = c("TGTAAATA", "TGCAT", "TTTTTTA", "ATATTC")

#this is new karsten/Chan et al 2018 data
  dr_raw <- read_tsv("data/new_karsten_dr_data.txt")
  codons_raw <- scan("data/codons.txt", character())
  
```

```{R clean data}
#create a dataframe with codon usage for each gene
ref_codons <- ref_raw %>%
  select(genename, CDS_seq,TTT:GGG) %>%
    rename(geneName = genename) %>%
    mutate_at(vars(TTT:GGG), funs(./nchar(CDS_seq)))




#change colnames, add means and filter missing values 
dr_raw2 <- dr_raw %>%
  rename_all(~c("geneName","gene","hlife_r1","hlife_r2")) %>%
    dplyr::select(-gene) %>%
      mutate(hlife = rowMeans(cbind(hlife_r1, hlife_r2), na.rm = TRUE)) %>% 
        filter(is.finite(hlife)) 

#cleaner df
dr_data <- dr_raw2 %>%
          dplyr::select(geneName, hlife)
  
```

Add codon usage to each dataset

```{R Add Codon Information}

#add dr data for Chan et al
ref_codons_dr <- left_join(ref_codons, dr_data, by = "geneName") %>%
  filter(is.finite(hlife))

```

```{R Multiple Linear Modelling}

#make model data
model_data <- ref_codons_dr %>%
  #convert to hlife to log2fold
    mutate(log2_hlife = log2(hlife)) %>%
      select(-hlife, -CDS_seq)

#create multiple lm model for prediction 
codons <- colnames(subset(model_data, 
                          select = -c(geneName, log2_hlife) ))

lm_codons_dr <- lm(data = model_data, 
                     paste("log2_hlife", 
                           paste(codons, 
                                 collapse = " + "),
                           sep = " ~ "))
```

Code for checking if sun et al codon usage values are correct
  (Take may take a couple minutes to run)
  
```{R Alternative Codon Count, eval=FALSE, include=FALSE}
#alternative method (testing if Sun's codon number are correct)
ref_codons2 <- tibble(genename = ref_raw$genename)


new_func <- function(seq_vec, codon_x = "TTC"){
  
  sst <- strsplit(seq_vec, "")
  
  sst2 <- lapply(sst, function(sst){
    do.call(paste0,
            lapply(seq_len(3), 
              function(i) {
                idx <- rep(FALSE, 3);
                idx[i] <- TRUE;
                sst[idx]
              })
            )
    })
  
  #search 
  sapply(sst2, function(ii){
    sum(str_count(ii, codon_x))
  })
    
}


```


Following code for using codon usage model to predict hlife is still incomplete 
```{R Prediction, eval=FALSE, include=FALSE}
predict(lm_motifs_k_dr, selected_data)
predicted_data <- model_data_sel %>%
  mutate(pred_log2hlife = predict(lm_motifs_k_dr, selected_data))

predicted_data %>%
  ggplot(aes(x = log2_hlife, y = pred_log2hlife)) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 4)) + 
    scale_y_continuous(limits = c(0, 4)) + 
    geom_smooth(method = "lm", formula = y ~ x) +
#    theme_coding() + 
    labs(y = "Predicted Half life", 
         x = "Measured Half Life")

```