---
title: "Codon Usage"
author: "Abhishek Jain"
date: "28th June 2019"
output: 
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
---

## Initation:

```{R Setup}
## Data exploration

if(!require(tidyverse)){
    install.packages("tidyverse")
    library(tidyverse)
}
if(!require(glmnet)){
    install.packages("glmnet")
    library(glmnet)
}
if(!require(lmodel2)){
    install.packages("lmodel2")
    library(lmodel2)
}
if(!require(ggplot2)){
    install.packages("ggplot2")
    library(tidyverse)
}

#theme for ggplot  
  theme_coding <- function(...){
  theme_bw()+
  theme(axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size = 12),
        panel.grid = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), units = , "cm"),
        plot.title = element_text(size = 20, vjust = 1, hjust = 0.5),
        legend.text = element_text(size = 12),
        legend.title = element_blank(),
        legend.position = c(0.9, 0.9))
}

#titles function incomplete
  x_y_titles <- function(){
    
    
  }
 
#defining fun color for scheming
colour <- c(teal = "#237E8A", green = "#1A8A21", red = "#A62621")

```

Important: DR Data used here is karsten weiss only.

```{R Load Data}

#get reference UTRs and CDS sequences from Cheng et al Suppl.
ref_raw <- read_rds("data/Sun_mutation_UTRs.rds")
  #Get sequences from ref_raw in a separate vector
  cds_seq <- ref_raw$CDS_seq
  
#Load Manually created motifs list into a vector
motifs_raw <- scan("data/list_motifs.txt", character())

#load list of 12 Chimeric Terminator Genes  
list_genes <- scan("data/genes_terminators_list.txt", character())

#manual Cheng motifs list
motifs_cheng = c("TGTAAATA", "TGCAT", "TTTTTTA", "ATATTC")

#this is new karsten/Chan et al 2018 data
  dr_raw <- read_tsv("data/new_karsten_dr_data.txt")

#For codon usage, get a list of codons
  codons_raw <- scan("data/codons.txt", character())
  
#single cell variability data from Nadal Ribelles et al   
  #sc_raw <-read_tsv("data/YPD_scRNA.tab")
```

```{R clean data}

#create a vector of 3'UTRs for searching motifs later 
UTR_3 <- ref_raw$UTR3_seq

#create a dataframe with codon usage for each gene
ref_codons_freq <- ref_raw %>%
  select(genename, CDS_seq, TTT:GGG) %>%
    rename(geneName = genename)

#Calculate proportional codon usage for each transcript & remove one 
#codon accounting for collinearity
  #NOTE: the order of codons (columns) gets flipped
ref_codons <- ref_codons_freq %>%
  gather(key = "codon", value = "freq", TTT:GGG) %>%
    mutate(freq = freq/nchar(CDS_seq)) %>%
  spread(codon, freq) %>%
    select(-TTT, -CDS_seq) 

codons <- colnames(ref_codons[-1])

#change colnames, add means and filter missing values 
dr_raw2 <- dr_raw %>%
  rename_all(~c("geneName","comgeneName","hlife_r1","hlife_r2")) %>%
    mutate(hlife = rowMeans(cbind(hlife_r1, hlife_r2), na.rm = TRUE)) %>% 
      filter(is.finite(hlife)) 

#cleaner df
dr_data <- dr_raw2 %>%
          dplyr::select(geneName, comgeneName, hlife)
  
```

Calculate 3'UTR Motif Frequencies from 3'UTR Ref sequences
```{R motif search}

#Dictionary for non-specific codes and converting U -> T
motifs <- motifs_raw %>% str_replace_all(c("U" = "T", "W" = "(A|T)", "S" = "(C|G)", "M" = "(A|C)", "K" = "(G|T)", "R" = "(A|G)", "Y" = "(C|T)", "B" = "(C|G|T)", "D" = "(A|G|T)", "H" = "(A|C|T)", "V" = "(A|C|G)", "N" = "(A|C|G|T)"))

#Initate ref tibble and store gene names
ref_motifs_raw <- tibble(geneName = ref_raw$genename)


#Search and add frequency of each c(motif) as a column in ref dataset
for (i in 1:length(motifs)){
ref_motifs_raw <- mutate(.data = ref_motifs_raw, 
                      !!motifs_raw[i] := str_count(UTR_3, motifs[i]))
}

motifs_present <- motifs_raw[sapply(ref_motifs_raw[-1], sum) > 0]

ref_motifs <-  ref_motifs_raw %>% 
  select_at(vars(geneName, motifs_present))
  

```

combine 
```{R Create model datasets}

#model df containing motifs and log2(hlife)
model_motifs <- left_join(ref_motifs, dr_data, by = "geneName") %>% 
    filter(is.finite(hlife)) %>% 
      #log transform hlife and remove
      mutate(log2_hlife = log2(hlife)) %>% select(-hlife) %>%
        #reorder columns
        select(geneName, comgeneName, log2_hlife, everything()) %>%
            #convert motifs data into binary
            mutate_at(vars(motifs_present), list(~ replace(., .>= 1, 1) )) %>%
            mutate_at(vars(motifs_present), list(~ factor(.) ))

#model df containing codon usage and and log2(hlife)
model_codons <- left_join(ref_codons, dr_data, by = "geneName") %>% 
    filter(is.finite(hlife)) %>% 
      #log transform hlife and remove original
      mutate(log2_hlife = log2(hlife)) %>% select(-hlife) %>%
        #reorder columns
        select(geneName, comgeneName, log2_hlife, everything())

#combine the two models
model_combined <- left_join(model_motifs, model_codons, 
                            by = c("geneName", "comgeneName", "log2_hlife"))

#model for selected genes
model_sel_genes <- model_combined %>%
  filter(geneName %in% list_genes)

#Create a subset of model without selected genes 

  #creates an operator for Not %in%
  '%!in%' <- function(x,y)!('%in%'(x,y))
  
    model_com_sub <- model_combined %>%
      filter(geneName %!in% list_genes)
    
```


```{R Linear Modelling}

predictors <- paste(c(paste(codons, collapse = " + "),
                  paste(motifs_present, collapse = " + ")), 
                collapse = " + ")

lm_codons_dr <- model_codons %>%
    filter(geneName %!in% list_genes) %>%
      select(-geneName, -comgeneName) %>%
        lm(data = ., paste("log2_hlife", 
                           paste(codons, 
                                 collapse = " + "),
                           sep = " ~ "))


lm_motifs_dr <- model_motifs %>%
    filter(geneName %!in% list_genes) %>%
      select(-geneName, -comgeneName) %>%
        lm(data = ., paste("log2_hlife", 
                           paste(motifs_present, 
                                 collapse = " + "),
                           sep = " ~ "))

lm_combined_dr <- lm(data = model_com_sub, paste("log2_hlife", predictors, sep = " ~ "))

lm_step <- step(lm_combined_dr, 
                scope = list(upper = lm_combined_dr, lower = lm_codons_dr),
                trace = FALSE)

#Create a list of motifs selected by step_lm
motifs_sel <- str_sub(names(coef(lm_step))[nchar(names(coef(lm_step))) > 3][-1], end = -2) 
```

```{R Prediction}
predicted_step_sel <- model_sel_genes %>%
  mutate(pred_hlife = predict(lm_step, 
                              model_sel_genes))

plot_predictions <- predicted_step_sel %>%
  ggplot(aes(x = log2_hlife, y = pred_hlife, col = paste(geneName, " (", comgeneName, ")"))) +
    geom_point() + 
    scale_x_continuous(limits = c(0, 5)) + 
    scale_y_continuous(limits = c(0, 5)) + 
    geom_abline() +
#    theme_coding() + 
    labs(y = "Predicted log2 (Hlife)", 
         x = "Measured log2 (Hlife)",
         colour = "geneID (geneName)",
         title = "Predicted vs Measured Half-life for 10 genes")

#display the plot
plot_predictions
```

```{R Test2}

predicted_step_sel %>% 
  select(geneName, comgeneName, log2_hlife, pred_hlife, motifs_sel) %>%
    mutate_if(function(col) {is.factor(col)}, strtoi) %>%
      select(function(col) {is.character})

predicted_step_sel %>% 
  select_if(is.character) %>%
    select(starts_with("Z")) %>%
  names() %>%
  mutate_at(df, ., paste, "b")

predicted_step_sel %>% 
  select(geneName, comgeneName, log2_hlife, pred_hlife, motifs_sel) %>%
    mutate_at(vars(motifs_sel), strtoi) %>% 
      select_if(function(col) {is.numeric(col) | is.character(col)})


predicted_step_sel %>% 
  mutate_at(vars(motifs_sel), strtoi) %>%
    select(motifs_sel) %>%
     
      select_if(colSums(.) > 0) %>%
        
x %>% select(c(which(sapply(., is.numeric)), "V4", "V5"))
starwars %>%
  select("name", names(toKeep)[as.numeric(toKeep) == 1])

```