---
title: "Abhi Motif Analaysis from Cheng et al 2017"
author: "Abhishek Jain"
date: "27th February 2019"
output: 
  md_document:
    variant: markdown_github  
editor_options: 
  chunk_output_type: console
---

## Objectives: 

  1. To create a script that calculates frequency of four specific motifs in yeast 3'UTR of all yeast gene (for each ORF)

  2. Find synthesis and degradation rates of each gene from Sun et al. data and add it to motif frequency table for further analysis
  

Method:
First, use cheng et al dataset (and previous dataframe codes) to create a table with wild type gene synthesis and decay rates

## Initation:

```{R Setup}
library(tidyverse)
library(ggplot2)
```




```{R Load Data}
#Error Handling for File not found
#Syntax tryCatch( {Exprns}, error = function(any){ print("out") } )
tryCatch({
            UTR_raw <- read_file("data/SGD_all_ORFs_3prime_UTRs.fsa")
            dr_raw <-read_tsv("data/sun_total_dr.txt")
            #sr_raw <- read_table("data/sun_total_sr.txt")
          },
        error = function(x) 
          { 
            print(x)
            cat("\nCurrent directory:",
                getwd())  
          }
)

#split single string into a vector to separate ORFs
  #[[1]] converts list of vectors to single vector 
UTR <- str_split(UTR_raw, ">sacCer3_ct_UserTrack_3545_")[[1]]

  #remove first empty element
  UTR <- tail(UTR, -1)

#change weird name of third column of wt degradation rate (to d_rate)
colnames(dr_raw)[3] <- "d_rate"

#new smaller dataframe with first three cols
dr_data <- dr_raw %>% 
  select(orf, gene, d_rate)
    
  
```

Then, create a tibble with motifs for every gene

```{R Freq of Motifs}

#vector containing end location of every ORF name from ref UTRs
pos = str_locate(UTR, "_")[2] - 1

#new dataframe with ORF names, and frequency counts of some motifs  
ref_motifs <- tibble(position = pos, orf = str_sub(UTR, 1, position), ATATTC = str_count(UTR, "ATATTAC"), TGCAT = str_count(UTR, "TGCAT"), TGTAAATA = str_count(UTR, "TGTAAATA"), TTTTTTA = str_count(UTR, "ATATTAC")) 

#delete the postiiton column: Not required
ref_motifs$position <- NULL

```



```{R Shrink and add rates}
#create a filtered dataset with consolidated ORFs
ref_motifs_fil <- ref_motifs %>%
    arrange(orf)%>%
    group_by(orf)%>%
      #keep max frequency of each motif 
      summarise_all(funs(max))%>%
    ungroup()%>%
    #add decay rate by matching orf name to dr dataframe
    mutate(dr = dr_data$d_rate[match(orf, dr_data$orf)])


```




```{R Next Chunk}


```



##Notes and Issues: 

Genes from Cheng et al have different names than SGD.  "sacCer3_ct_UserTrack_3545_YDR461W_SC_Transcript_00013696_Miura_2006 range=chrIV:1385287-1385434 5'pad=0 3'pad=0 strand=+ repeatMasking=none" 


1. Number of Genes is very different: 

    From SGD Data, Only 2827 obs. left (of 17903 initiatlly) after deduplicating ORFs of same genes [Taking max number of each motifs in each case]. 
    Cheng et al. data had 4305 genes for synthesis and degradation rates among all mutants.



