---
title: "abhi_rmd_v1"
author: "Abhishek Jain"
date: "08th February 2019"
output: 
  md_document:
    variant: markdown_github  
editor_options: 
  chunk_output_type: console
---


## Initation

```{R Setup}
library(tidyverse)
library(ggplot2)
```

## Task 1.1: Load Datasets

Loading the datasets dr_data.csv and sr_data.csv in objects by same names.

------ SAM NOTES -----

Be careful! You have used a read.csv function here which is a base function (i.e. it come with R as standard). However this task is about using tidyverse, so you should use the equivalent tidyverse function read_csv!

What is the difference? read.csv will create a dataframe to hold the imported information, meanwhile read_csv will create a tibble to hold the imported data.

To understand the important differences between a tibble and a dataframe type this into the console, vignette("tibble")

They are subtly different but little differences can have huge effects! If you are using tidyverse you need to make sure all the functions you use accept tibbles as inputs and creates tibbles as outputs! (tidyverse is a group of functions with this exact purpose)

Im impressed that you looked into error handling! Very useful if you are creating software to be distributed.

```{r Load Data}
#Error Handling for File not found
#Syntax tryCatch( {Exprns}, error = function(any){ print("out") } )
tryCatch({
            dr_data <- read_csv("data/dr_data.csv")
            sr_data <- read_csv("data/sr_data.csv")
          },
        error = function(x) 
          {
            cat("\nCheck Working Directory. Current directory:\n",
                getwd())
          }
)
```

------ SAM NOTES -----
Again the merge function is base function, therefore outputs dataframes not tibbles. Try left_join instead.

```{R Combine SR and DR}
#new dataframe containing both data together by Gene and Strain, 
#Suffixes dr and sr for distinguishing values.
combined_data <- merge(dr_data, sr_data,
                       by = c("Gene", "X1", "Strain"), 
                       suffixes = c("_dr", "_sr"))
```

##Task 1.2: Calculate Mean SR and DR

Calculating the mean of decay (drm) and synthesis (srm) rates for each gene across all strains containing it. Unit is: Log2fold

------ SAM NOTES -----
Very happy that you are getting into commenting but there is such a thing as over commenting.

Try not to put comments inbetween pipes operators. Comments are there to provide contextual information, not to simply repeat the command, i.e. you dont need to explain the function group_by(gene) by saying group rows by gene. 

The point of piping is to make complex manipulations easily readable. If you are calling functions with already well written documentation you do not need to comment what it is doing. Only explain usage of user defined functions.

Also try not to comment out sections of the pipe (either have it in or dont!).

Something like this is fine

#Calculate each genes mean rate, removing any poor data
com_data_m <- combined_data %>%
  distinct() %>%
    filter(is.finite(Log2Fold_dr), is.finite(Log2Fold_sr)) %>%
      group_by(Gene) %>%
        mutate(drm = mean(Log2Fold_dr), srm = mean(Log2Fold_sr)) %>%
          ungroup()
        
```{r Calculating Mean DR and SR}
com_data_m <- combined_data %>%
  #Remove duplicate rows
  distinct() %>%
    #remove NAs in both log2fold column
    #filter(is.finite(Log2Fold_dr), is.finite(Log2Fold_sr)) %>%
      # Group rows so that each group is one gene
      group_by(Gene) %>%
        #calculate mean
        mutate(drm = mean(Log2Fold_dr), srm = mean(Log2Fold_sr))%>%
      ungroup()  
```

## Task 1.3: Find Significant Changes

Create SINGLE table of genes that have one or more mutants with significantly changed decay rate OR synthesis rate? 
Also, record the number of significant mutants for each gene.

------ SAM NOTES -----

Same as above

```{R Significance2}
#Analyze this dataset
sig_data <- com_data_m %>% 
  arrange(Gene)%>%
  #select(Gene, Significance_dr, Significance_sr)%>%
    filter(Significance_dr != "not" | Significance_sr != "not")%>%
      group_by(Gene)%>%
        #calculate and add number of sig mutants for every gene
        mutate(No_of_sig_mutants = n())%>%
          #Remove rows with no difference in significance between genes
          #distinct()%>%
      ungroup()
               
```


##Task 1.4: Log2Fold Readings in Gene vs Mutant Table 

Goal: To Make another table with one row per gene; with the logfold decay/synthesis rates across all mutants along the columns

------ SAM NOTES -----

Same as above, also why are you grouping by gene?

```{R Tranposing Decay Table}
#Analyze this dataset
trans_data_dr <- com_data_m %>% 
    #get rid of other columns
    select(Gene, Strain, Log2Fold_dr, drm)%>%

        #Transposing (Key,Value) 
        #Strains go to columns (Key); Log2Fold to row (values)
        spread(Strain, Log2Fold_dr)%>%

      ungroup()%>%
    arrange(desc(drm)) #just for better handling

```

Same for Synthesis
```{R Tranposing Synthesis Table}
#Analyze this dataset
trans_data_sr <- com_data_m %>% 
    #get rid of other columns
    select(Gene, Strain, Log2Fold_sr, srm)%>%
      #Grouping
      group_by(Gene)%>%
        
        #Transposing (Key,Value) 
        #Strains go to columns (Key); Log2Fold to row (values)
        spread(Strain, Log2Fold_sr)%>%

      ungroup()%>%
    arrange(desc(srm)) #just for better handling

```



##Task 1.5: Plotting

---- SAM NOTES ----

Instead of using arrange and head functions you could use top_n function

#Select top 5 genes with heighest mean decay rates
top_data_dr <- trans_data_dr %>% 
      top_n(5,drm)


```{R Create Data for Plot}
#Start with dr data with mean
top_data_dr <- trans_data_dr %>% 
    #get rid of other columns
      #arrange top 5 genes by mean dr
      arrange(desc(drm))%>%
      #Take top 5
      head(n = 5)

#Start with sr data with mean
top_data_sr <- trans_data_sr %>% 
      #arrange top 5 genes by mean sr
      arrange(desc(srm))%>%
      #Take top 5
      head(n = 5)

#Find the top5 Genes in actual data
com_data_plot <- com_data_m %>%

  filter(Gene %in% top_data_dr$Gene | Gene %in% top_data_sr$Gene)%>%
    select(Gene, Strain, Log2Fold_dr, Log2Fold_sr)

plot_data <- gather(com_data_plot, key = "Type", value = "Log2fold", -Gene, -Strain)

```

```{R ggPlot Synthesis Rate vs Genes}
#Create the plot
(my_plot <- ggplot(plot_data, aes(x = Log2fold, y = Strain, colour = Gene, shape = Type)) +
   geom_point()
)
```

Seems to work although the number of points is too high. 

ISSUES:
  com_data_plot currently filters for all Top genes in either decay or synthesis rate while we only want the top ones from each of them. 
  



## Task 2: Regex

## Task 2.1: Import Data
---- SAM NOTES ----

Again fun to mess around with error messages but do not need to this every time! Not exactly useful because you are guessing what could be wrong not telling the user what IS wrong.
```{R Import Chromosome 7}
#Error Handling for File not found
#Syntax tryCatch( {Exprns}, error = function(any){ print("out") } )
tryCatch({
            chr7 <- read_file("data/ScerevisiaeChromVIII.txt")
            #PS File name is ChromVIII but the data is for ChromVII
          },
        error = function(x) 
          { 
            print(x)
            cat("\nCurrent directory:",
                getwd())  
          }
)

```

## Task 2.2: Counting frequency of "TGTTGGAATA"

Counting the number of occurrences of TGTTGGAATA in chr7 using stringr::str_count  
```{R Look for TGTTGGAATA}
print(str_count(chr7, "TGTTGGAATA"))
```

## Task 2.3: Finding Most common bases before and after "TGTTGGAATA" motifs
```{R Finding Common Bases Before/After}
#create a df of start/stop indices for ".XXXX." motif
motif_all <- data.frame(str_locate_all(chr7, ".TGTTGGAATA."))

#Use the df for identifying bases
motif_all <- motif_all %>%
  mutate(
    #identify the base before and after
    before = str_sub(chr7, start, start), 
    after = str_sub(chr7, end, end),
    #JFF - Find the actual expanded sequence with random base 
    exp_seq = str_sub(chr7, start, end)
    )
  
#Print Most common mutants using max() and min()
max(motif_all$before)
max(motif_all$after)


```


##Task 2.4: Finding single nt Mutants for "TGTTGGAATA


```{R Mutants of TGTTGGAATA}
motif <- "TGTTGGAATA"
mutants <- c()

#A loop for creating a list of candidate mutants to look for 
for (base in 1:str_length("TGTTGGAATA")) {
  #substitute the base with current loop index with wildcard(.)
  str_sub(motif, base, base) <- "."
  #add to list
  mutants[[base]] <- motif  
  #reset the motif for next iteration
  motif <- "TGTTGGAATA"
}

#print(mutants)

#find mutants and store as a list of matrices
x <- str_locate_all(chr7, mutants)
#convert the list of matrices to a dataframe 
chr7_mutants <- data.frame(do.call(rbind, x))

# Filter out duplicates and Identify the mutant motifs   
chr7_mutant_motifs <- chr7_mutants %>%
    mutate(motif = str_sub(chr7, start, end))%>%
      #only keep sequence motifs
      select(motif)%>%
        #remove duplicate motifs
        unique()%>%
          #filter out original motif 
          filter(motif != "TGTTGGAATA")
        
      
print(chr7_mutant_motifs)

```